### PR TITLE
SslHandler flushing with TCP Fast Open fix

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
SslHandler owns the responsibility to flush non-application data
(e.g. handshake, renegotiation, etc.) to the socket. However when
TCP Fast Open is supported but the client_hello cannot be written
in the SYN the client_hello may not always be flushed. SslHandler
may not wrap/flush previously written/flushed data in the event
it was not able to be wrapped due to NEED_UNWRAP state being
encountered in wrap (e.g. peer initiated renegotiation).

Modifications:
- SslHandler to flush in channelActive() if TFO is enabled and
  the client_hello cannot be written in the SYN.
- SslHandler to wrap application data after non-application data
  wrap and handshake status is FINISHED.
- SocketSslEchoTest only flushes when writes are done, and waits
  for the handshake to complete before writing.

Result:
SslHandler flushes handshake data for TFO, and previously flushed
application data after peer initiated renegotiation finishes.

### CODE REVIEW

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketSslEchoTest.java [hunk 1]
```diff
@@ -265,6 +267,7 @@ public void initChannel(Channel sch) throws Exception {
                 } else {
                     serverSslHandler = serverCtx.newHandler(sch.alloc());
                 }
+                serverSslHandler.setHandshakeTimeoutMillis(0);
```
Comment (Scottmitch) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketSslEchoTest.java:270]: the motivation here is to rely upon the test timeout, makes debugging easier if there is a single timeout for the test but this isn't necessary.

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 2]
```diff
@@ -1967,6 +1968,9 @@ public void handlerAdded(final ChannelHandlerContext ctx) throws Exception {
             // With TCP Fast Open, we write to the outbound buffer before the TCP connect is established.
             // The buffer will then be flushed as part of establishing the connection, saving us a round-trip.
             startHandshakeProcessing(active);
+            // If we weren't able to include client_hello in the TCP SYN (e.g. no token, disabled at the OS) we have to
+            // flush pending data in the outbound buffer later in channelActive().
+            needsFlush |= fastOpen && channel.unsafe().outboundBuffer().totalPendingWriteBytes() > 0;
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java:1973]: `channel.unsafe().outboundBuffer()` may return `null` when a channel was closed we should add a null check.

```suggestion
            ChannelOutboundBuffer outboundBuffer = channel.unsafe().outboundBuffer();
            needsFlush |= fastOpen && outboundBuffer != null && outboundBuffer.totalPendingWriteBytes() > 0;
```

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 3]
```diff
@@ -1790,40 +1796,35 @@ public void run() {
...
                         session.getProtocol(),
                         session.getCipherSuite());
             }
             ctx.fireUserEventTriggered(SslHandshakeCompletionEvent.SUCCESS);
-        }
 
-        if (readDuringHandshake && !ctx.channel().config().isAutoRead()) {
-            readDuringHandshake = false;
-            ctx.read();
+            if (readDuringHandshake && !ctx.channel().config().isAutoRead()) {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java:1821]: I would keep this out of the branch (just like before).

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: handler/src/test/java/io/netty/handler/ssl/RenegotiateTest.java [hunk 4]
```diff
@@ -55,7 +55,9 @@ public void testRenegotiateServer() throws Throwable {
                     .childHandler(new ChannelInitializer<Channel>() {
                         @Override
                         protected void initChannel(Channel ch) throws Exception {
-                            ch.pipeline().addLast(context.newHandler(ch.alloc()));
+                            SslHandler handler = context.newHandler(ch.alloc());
+                            handler.setHandshakeTimeoutMillis(0);
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/RenegotiateTest.java:59]: why this change ?

**Comment by Scottmitch** [comment]
build failure attributed to https://github.com/netty/netty/issues/11080. seems unrelated to this PR as these tests don't use SslHandler.

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 3 above)
Comment (Scottmitch) [reply handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: done

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 2 above)
Comment (Scottmitch) [reply handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: good catch! I will also flush if `outboundBuffer == null` (better to do an extra flush rather than miss a flush).

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 5]
```diff
@@ -1790,40 +1797,34 @@ public void run() {
...
-        boolean notified = handshakePromise.trySuccess(ctx.channel());
-        SSLSession session = engine.getSession();
-
-        // There seems to be a bug in the SSLEngineImpl that is part of the OpenJDK that results in returning
-        // HandshakeStatus.FINISHED multiple times which is not expected. This only happens in TLSv1.3 so lets
-        // ensure we only notify once in this case.
-        //
-        // This is safe as TLSv1.3 does not support renegotiation and so we should never see two handshake events.
-        if (notified || !SslUtils.PROTOCOL_TLS_V1_3.equals(session.getProtocol())) {
+    private boolean setHandshakeSuccess() {
```
Comment (chrisvest) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java:1806]: Would be good to also add a javadoc comment about the new return values.

**Review by chrisvest — APPROVED** [review]
Had a small comment about javadoc, but otherwise it looks good to me.

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 5 above)
Comment (Scottmitch) [reply handler/src/main/java/io/netty/handler/ssl/SslHandler.java:1808]: done

**Review by normanmaurer — APPROVED** [review]
